services/gitdiff/gitdiff.go

```diff
@@ -555,6 +283,182 @@ func DiffInlineWithHighlightCode(fileName, language, code string) DiffInline {
 	return DiffInline{EscapeStatus: status, Content: template.HTML(content)}
 }
 
+// HighlightCodeDiff is used to do diff with highlighted HTML code.
+// The HTML tags will be replaced by Unicode placeholders: "<span>{TEXT}</span>" => "\uE000{TEXT}\uE001"
+// These Unicode placeholders are friendly to the diff.
+// Then after diff, the placeholders in diff result will be recovered to the HTML tags.
+// It's guaranteed that the tags in final diff result are paired correctly.
+type HighlightCodeDiff struct {
+	placeholderBegin    rune
+	placeholderMaxCount int
+	placeholderIndex    int
+	placeholderTagMap   map[rune]string
+	tagPlaceholderMap   map[string]rune
+
+	lineWrapperTags []string
+}
+
+func NewHighlightCodeDiff() *HighlightCodeDiff {
+	return &HighlightCodeDiff{
+		placeholderBegin:    rune(0xE000), // Private Use Unicode: U+E000..U+F8FF, BMP(0), 6400
+		placeholderMaxCount: 6400,
```

Comment:
No need to use that large.
As https://github.com/go-gitea/gitea/pull/19958#discussion_r899948331

Reply:
I am aware that only the chroma tags are what are being replaced - I'm also aware that we shouldn't be running diff on a line with more than `MAX_GIT_DIFF_LINE_CHARACTERS` (5000 characters).

However, consider:

`a = "`+ a string of 3200 supplementary characters + `"` 

diffed with: 

`b = "" // `+ a string of 3200 different supplementary characters + `"` 

so long as all 6400 of the supplementary characters are included no replacements can be made. The bits before and within the supplementary characters can be tuned to cause HTML breakage. I don't think it would be easy to get arbitrary tag creation but certainly HTML breakage is possible.

Reply:
I do no think the HTML breakage is serious in this case. It will be corrected by browsers in the end by the surrendering `<div>` tags. Chroma only outputs `<span>` tags. Browsers can handle it correctly. If you want to make sure there is no breakage at all, the final HTML can be sanitized for tag-matching again.

Reply:
And at the moment, the code in `recoverOneDiff` uses a stack to output final tags, all tags should have been paired correctly when outputing, I do not think it would output broken HTML tags.

Reply:
I've just shown how you can get to 6400 supplementary characters even within our current defaults.

My comment regarding broken tags was before I noted you'd changed lines 414-415 to simply not emit the tag which helps prevent the broken code however, we do need to consider the issue of mismatching...

Consider the above example but one that only covers 6399 supplementary characters, it will not be possible to create a placeholder for the closing tag so its closing tag will not get covered. So we're guaranteed in that case to end up with a mismatched tag.

It would be a simple case to assign the closing placeholder at the same time as the opening placeholder to prevent this.

Reply:
> So we're guaranteed in that case to end up with a mismatched tag.

That's impossible, the tags are closed by the stack check.

Reply:
ps: "It would be a simple case to assign the closing placeholder at the same time as the opening placeholder to prevent this."

Although it's feasible, but it doesn't help much, because the opening/closing tags could still mismatch during diff.

So in any case, the closing tags should be added by the stack. Then no need to worry about there is no placeholder for closing tags.